### PR TITLE
pRuntime: Refactor benchmark codes & play with pallet-registry.

### PR DESCRIPTION
**Pull request by kvinwang**
### Worker state machine
Note: The worker state machine is different from the original design, coz i consider it will be simpler that spliting the state into smaller ones. We may need more discussion on it.

### Bechmark score

The benchmark score in my PC:
|||
| --- | --- |
| CPU | Intel(R) Core(TM) i7-9700K CPU @ 3.60GH (8 cores) |
| SGX_MOD | SW |
| Bench cores | 8 |
| Score | 475 iter/s (60 iter/core/s) |

Need any futher tuning?

### Benchmark counter overflow
```python
In [2]: 0xffff_ffff_ffff_ffff / (475*3600*24*365)
Out[2]: 1231457720.74752
```
It will take 1231457720.74752 years to overflow our u64 counter variable for my PC.

### Deterministic
The benchmark score reported by pRuntime is not deterministic. Seems not a big problem here?

### CODE REVIEW

File: pallets/phala/src/phala_legacy/tests.rs [hunk 1]
```diff
@@ -163,19 +160,16 @@ fn test_register_worker() {
 		));
 		assert_ok!(Phala::set_stash(Origin::signed(1), 1));
 		// TODO(jasl): fix runtime info
-		assert_ok!(Phala::register_worker(
+		// TODO: we need to re-design the tests.
```
Comment (h4x3rotab) [inline pallets/phala/src/phala_legacy/tests.rs:163]: We will will retire phala_legacy. So we can feel free to delete the out-of-date tests

File: pallets/phala/src/registry.rs [hunk 2]
```diff
@@ -166,23 +183,34 @@ pub mod pallet {
 					Some(worker_info) => {
 						// Case 1 - Only refresh the RA report; no need to redo benchmark
```
Comment (h4x3rotab) [inline pallets/phala/src/registry.rs:184]:
```suggestion
						// Case 1 - Refresh the RA report and redo benchmark
```

File: standalone/phost/src/main.rs [hunk 3]
```diff
@@ -582,7 +575,15 @@ async fn bridge(args: Args) -> Result<()> {
...
             warn!("pRuntime not initialized. Requesting init...");
             runtime_info = Some(
-                init_runtime(&client, &pr, !args.ra, args.use_dev_key, &args.inject_key).await?,
+                init_runtime(
+                    &client,
+                    &pr,
+                    !args.ra,
+                    args.use_dev_key,
+                    &args.inject_key,
+                    args.operator.clone(),
```
Comment (h4x3rotab) [inline standalone/phost/src/main.rs:584]: Let's use [`from_str`](https://substrate.dev/rustdocs/v3.0.0-monthly-2021-07/src/sp_core/crypto.rs.html#823-837) to allow ss58 encoded address:

```rust
let parsed_operator = sp_core::crypto::AccountId32::from_str(&args.operator)?;
let operator_hex = hex::encode(&parsed_operator);
```

**Review by h4x3rotab — COMMENTED** [review]
Btw, should we pause benchmark when it's not needed (after register_worker benchmark)? No need to implement it in this PR. We can leave a TODO if you think it makes sense.

**Comment by h4x3rotab** [comment]
Did you do the benchmark in `SGX_MODE=SW`?

File: pallets/phala/src/phala_legacy/tests.rs (hunk 1 above)
Comment (kvinwang) [reply pallets/phala/src/phala_legacy/tests.rs:163]: So I removed the whole test file. If we need it as a referance when we write new tests, we can recover it from git.

File: pallets/phala/src/registry.rs [hunk 4]
```diff
@@ -227,19 +255,42 @@ pub mod pallet {
...
+				RegistryEvent::BenchReport {
+					start_time,
+					iterations,
+				} => {
+					let now = T::UnixTime::now().as_secs().saturated_into::<u64>();
+					if now <= start_time {
+						// Oops, should not happen
+						return Err(Error::<T>::InvalidBenchReport.into());
+					}
+					let score = iterations / (now - start_time);
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply pallets/phala/src/registry.rs:278]: Done

**Review by h4x3rotab — APPROVED** [review]

**Pull request merged**